core/types/subtyping.cc

```diff
@@ -404,28 +404,35 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
             ENFORCE(result != nullptr);
             return result;
         } else {
-            bool allowProxyInLub = isa_type<TupleType>(t1) || isa_type<ShapeType>(t1);
             // only 1st is proxy
-            TypePtr und = t1.underlying(gs);
-            if (isSubType(gs, und, t2)) {
-                return t2;
-            } else if (allowProxyInLub) {
-                return OrType::make_shared(t1, t2);
-            } else {
-                return lub(gs, t2, und);
-            }
+            return lub(gs, t2, t1);
```

Comment:
This won't work, because we already reverse/recurse on `t1.kind() > t2.kind()`. https://github.com/sorbet/sorbet/pull/4495#discussion_r689782209